### PR TITLE
Orders: Add bulk action to refund overpaid amount

### PR DESCRIPTION
**Pull request by jeriox**
closes #3699

### CODE REVIEW

**Comment by jeriox** [comment]
@raphaelm as promised in the discussion, I started to work on a PR :) this is a first implementation, but it already works for me. At this point, I have two questions:

- The annotation for the queryset is copied over from the one in the order list. What would be an appropriate spot to put it so that both views can use it?
- The current code allows all orders that are overpaid to be selected for the bulk action. However, there will be orders that are overpaid but cannot be refunded automatically. In the processing of a single order, this will be catched by `propose_auto_refunds()`. As far as I see, there is no possibility to mark the operation as failed for single orders. The UI would therefore currently just display a success message. For me that would be fine as the orders that could not be refunded would still be marked as overpaid. Just filtering out the manual payment provider in the qs is probably not enough, as there are other conditions where payment providers would mark the payment as not refundable. What are your thoughts on this?

File: src/pretix/control/views/orders.py [hunk 1]
```diff
@@ -316,6 +317,54 @@ def execute_single(self, instance, form: forms.Form):
...
+                refund = OrderRefund.objects.create(
+                    order=instance,
+                    payment=payment,
+                    source=OrderRefund.REFUND_SOURCE_ADMIN,
+                    state=OrderRefund.REFUND_STATE_CREATED,
+                    amount=amount,
+                    comment=_("Refund for overpayment"),
+                    provider=payment.provider
+                )
+                payment.payment_provider.execute_refund(refund)
```
Comment (raphaelm) [inline src/pretix/control/views/orders.py:365]: This will need some error handling so the entire process is not stopped when a single refund fails

**Review by raphaelm — COMMENTED** [review]
> The annotation for the queryset is copied over from the one in the order list. What would be an appropriate spot to put it so that both views can use it?

There is already a classmethod called `Order.annotate_overpayments` that you should be able to use here (e.g. `Order.annotate_overpayments(queryset).filter(is_overpaid=True)`)

> The current code allows all orders that are overpaid to be selected for the bulk action. However, there will be orders that are overpaid but cannot be refunded automatically. In the processing of a single order, this will be catched by propose_auto_refunds(). As far as I see, there is no possibility to mark the operation as failed for single orders. The UI would therefore currently just display a success message. For me that would be fine as the orders that could not be refunded would still be marked as overpaid. 

I agree. Ideally, the success message would say something like "X orders have been refunded, Y still need manual attention".

**Comment by jeriox** [comment]
> There is already a classmethod called `Order.annotate_overpayments` that you should be able to use here (e.g. `Order.annotate_overpayments(queryset).filter(is_overpaid=True)`)

Thanks for the pointer!

> I agree. Ideally, the success message would say something like "X orders have been refunded, Y still need manual attention".

I moved the logic up to the base class, as counting there is easier. Tell me what you think about that, if you don't like that I'll just move the success message down to my action

**Review by raphaelm — COMMENTED** [review]
Code LGTM, but this is a change we'll need a signed CLA for.

**Comment by raphaelm** [comment]
@cla-bot[bot] check

**Comment by jeriox** [comment]
That was insanely fast :D 
will write a test in the next few days to make the other bot happy as well

**Comment by raphaelm** [comment]
I would merge this without the tests, but the unused imports need to be cleaned up, i.e.

```
commit 87e0829a867ce0321466938552bdf0e40a8a6f84 (HEAD -> overpaid-bulk-refund)
Author: Raphael Michel <michel@rami.io>
Date:   Wed Nov 22 15:17:19 2023 +0100

    Fix flake8 errors

diff --git a/src/pretix/control/views/orders.py b/src/pretix/control/views/orders.py
index a04390bb8..ef9a59732 100644
--- a/src/pretix/control/views/orders.py
+++ b/src/pretix/control/views/orders.py
@@ -47,12 +47,11 @@ from django.conf import settings
 from django.contrib import messages
 from django.core.exceptions import PermissionDenied, ValidationError
 from django.core.files import File
-from django.db import models, transaction
+from django.db import transaction
 from django.db.models import (
-    Case, Count, Exists, F, IntegerField, OuterRef, Prefetch, ProtectedError,
-    Q, QuerySet, Subquery, Sum, Value, When,
+    Count, Exists, F, IntegerField, OuterRef, Prefetch, ProtectedError, Q,
+    QuerySet, Subquery, Sum,
 )
-from django.db.models.functions import Coalesce
 from django.forms import formset_factory
 from django.http import (
     FileResponse, Http404, HttpResponseNotAllowed, HttpResponseRedirect,
```

I tried to commit it myself but GitHub wouldn't let me

File: src/pretix/control/views/orders.py [hunk 2]
```diff
@@ -314,6 +324,36 @@ def allowed_for(self, queryset):
...
         mark_order_expired(instance, user=self.request.user)
+        return True
+
+
+class OrderOverpaidRefundBulkActionView(BaseOrderBulkActionView):
+    label = _("Refund overpaid amount")
+
+    def allowed_for(self, queryset):
+        return Order.annotate_overpayments(queryset).filter(
+            status=Order.STATUS_PAID,
```
[the remark this answers is not in the document]
Comment (jeriox) [reply src/pretix/control/views/orders.py:335]: fair point, our use case involves only cancelled positions, that's why I came up with that in the first place

**Comment by jeriox** [comment]
> I tried to commit it myself but GitHub wouldn't let me

weird, "Allow edits by maintainers" is enabled

**Review by raphaelm — APPROVED** [review]

**Pull request merged**